csharp/Ice/async/Server.cs

```diff
@@ -12,51 +14,38 @@ public static int Main(string[] args)
 
         try
         {
-            //
-            // using statement - communicator is automatically destroyed
-            // at the end of this statement
-            //
-            using(var communicator = Ice.Util.initialize(ref args, "config.server"))
+            using (Ice.Communicator communicator = Ice.Util.initialize(ref args, "config.server"))
             {
-                if(args.Length > 0)
+                if (args.Length > 0)
                 {
-                    Console.Error.WriteLine("too many arguments");
+                    Console.WriteLine("too many arguments");
                     status = 1;
                 }
                 else
                 {
-                    var workQueue = new WorkQueue();
-
-                    //
-                    // Shutdown the communicator and destroy the workqueue on Ctrl+C or Ctrl+Break
-                    // (shutdown always with Cancel = true)
-                    //
-                    Console.CancelKeyPress += (sender, eventArgs) =>
+                    using (var cts = new CancellationTokenSource())
                     {
-                        eventArgs.Cancel = true;
-                        workQueue.destroy();
-                        communicator.shutdown();
-                    };
+                        Console.CancelKeyPress += (sender, eventArgs) =>
+                        {
+                            eventArgs.Cancel = true;
+                            cts.Cancel();
+                        };
 
-                    var adapter = communicator.createObjectAdapter("Hello");
-                    adapter.add(new HelloI(workQueue), Ice.Util.stringToIdentity("hello"));
-
-                    workQueue.Start();
-                    try
-                    {
+                        Ice.ObjectAdapter adapter = communicator.createObjectAdapter("Hello");
+                        adapter.add(new HelloI(cts), Ice.Util.stringToIdentity("hello"));
                         adapter.activate();
-                        communicator.waitForShutdown();
-                    }
-                    finally
-                    {
-                        workQueue.Join();
+
+                        // cts is canceled by Ctrl+C or a shutdown request.
+                        // In modern C#, you should call: await Task.Delay(-1, cts.Token)
+                        cts.Token.WaitHandle.WaitOne();
+                        communicator.shutdown();
```

Comment:
I would not say "in modern C#". Or at least say that we need to do it this way for compatibility with older versions of .NET

Reply:
Fixed